docs/design/resources.md

```diff
@@ -0,0 +1,96 @@
+# Operator Resources
+
+The etcd cluster operator extends the Kubernetes API with two new
+resource definitions:
+
+* `EtcdCluster`, or the "cluster resource", defines an entire
+  operating cluster.
+* `EtcdPeer`, or the "peer resource", defines a single peer of the
+  etcd cluster. This principally concerns a Pod, but may also involve
+  attached storage and network identity.
+
+## The Peer Resource
+
+The peer resource represents a single peer in an etcd
+cluster. Encoding it's bootstrap instructions, persistence settings,
+and extra pod settings (e.g., overridden CPU limits). Its `status`
+field exposes information about the peer including its hostname,
+liveness status, the peer's current generation, and if it's currently
+the leader of the etcd cluster.
+
+When reconciling the etcd peer resource the operator will do the
+following:
+
+1. If storage is enabled and a correctly named persistent volume claim
+   (PVC) does not exist, create one with the settings provided on the
+   peer resource.
+
+   We never delete a persistent volume claim automatically, so this
+   behaviour ensures that the operator will resume stopped clusters if
+   the storage remains even if the peer resource was deleted.
+
+   If the storage options on the peer resource are different from an
+   existing PVC, the operator makes no attempt to 'correct' the
+   preexisting PVC.
+2. If a correctly named pod does not exist, an etcd pod will be
+   launched with the PVC from above mounted as a volume if storage is
+   enabled. The bootstrap configuration will always be provided to the
+   pod as configuration, as etcd will ignore bootstrap configuration
+   if a data directory already exists.
+3. If the pod is in a ready state, the operator will attempt to
+   contact it using the gRPC endpoint and retrieve information to
+   populate the status field. Otherwise, it will report the peer as
+   not-ready on the status field.
```

Comment:
Let's roll with this for now, although I think we should also consider:

1) using readiness probes for at least part of this
2) how often we poll (and consequently how much network traffic is generated)
3) what guarantees we can realistic give about our API given this polling frequency

The EtcdCluster controller typically will typically work best when working with guaranteed-to-be-up-to-date data, as it avoids consistency issues between the k8s API/listers and etcd itself. For that reason, I think we'll end out *not* relying on the peer resources status fields too extensively when it comes to taking potentially dangerous cluster operations.

Reply:
Not that I think it's worth changing here, but just to note, we'll probably want to try probing the peer even if the *Pod* is not ready (but *is* 'alive', i.e. not restarting). The API may still be available, and may provide extra insight as to why it's in its current state.